### PR TITLE
[RFR] Api file import

### PR DESCRIPTION
**Pull request by ThieryMichel**

### CODE REVIEW

File: src/app/js/home/Home.js [hunk 1]
```diff
@@ -1,3 +1,6 @@
 import React, { PropTypes } from 'react';
+import Upload from './Upload';
```
Comment (djhi) [inline src/app/js/home/Home.js]: If #7 is merged first, you'll be able to put this Upload component on the Admin component

File: package.json [hunk 2]
```diff
@@ -20,21 +20,25 @@
   },
   "homepage": "https://github.com/Inist-CNRS/lodex-v2#readme",
   "dependencies": {
+    "async-busboy": "^0.3.3",
```
Comment (djhi) [inline package.json]: replace `^` with `~`

**Review by djhi — CHANGES_REQUESTED** [review]

File: makefile [hunk 3]
```diff
@@ -52,19 +55,24 @@ build-app: ## Build the frontend application
...
-docker-run-dev: ## run node server with pm2 for development and webpack-dev-server 
-	docker-compose up --force-recreate
 
 test-app-unit: ## Run the frontend application unit tests
 	NODE_ENV=test BABEL_ENV=browser ./node_modules/.bin/mocha \
 		--require='./src/app/js/test.spec.js' \
 		--compilers="css:./src/common/tests/webpack-null-compiler,js:babel-core/register" \
 		"./src/app/js/**/*.spec.js"
 
+<<<<<<< HEAD
```
Comment (djhi) [inline makefile]: oops

File: src/api/controller/api/index.js [hunk 4]
```diff
@@ -3,11 +3,24 @@ import route from 'koa-route';
...
+import upload from './upload';
 
 const app = new Koa();
 
+app.use(async (ctx, next) => {
+    ctx.db = await mongoClient();
+    try {
+        await next();
+    } finally {
+        console.log('closing db');
```
Comment (djhi) [inline src/api/controller/api/index.js]: debug spotted

**Review by djhi — CHANGES_REQUESTED** [review]

File: src/api/controller/api/index.js [hunk 5]
```diff
@@ -3,16 +3,31 @@ import route from 'koa-route';
...
+import mongoClient from '../../services/mongoClient';
 import login from './login';
+import upload from './upload';
+import dataset from '../../models/dataset';
+import parsingResult from '../../models/parsingResult';
+import parsing from './parsing';
 
 const app = new Koa();
 
+app.use(async (ctx, next) => {
```
Comment (djhi) [inline src/api/controller/api/index.js]: Should be put in a middleware file to allow testing. Not very important

File: src/api/models/dataset.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import chunk from 'lodash/chunk';
+
+export default (db) => {
+    const collection = db.collection('dataset');
+
+    collection.insertBatch = documents => chunk(documents, 100).map(data => collection.insertMany(data));
+
```
Comment (djhi) [inline src/api/models/dataset.js]: extra line

File: src/app/.eslintrc [hunk 7]
```diff
@@ -12,7 +12,6 @@
         "react/jsx-filename-extension": ["warn", { "extensions": [".js", ".jsx"] }],
```
Comment (djhi) [inline src/app/.eslintrc]: We don't need react on the API

File: src/app/js/admin/upload/Upload.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+import { bindActionCreators } from 'redux';
+import { connect } from 'react-redux';
+import RaisedButton from 'material-ui/RaisedButton';
+import ArchiveIcon from 'material-ui/svg-icons/content/archive';
+
+import { uploadFile } from './';
+
+export const UploadComponent = ({ onFileLoad }) => (
+    <div
+        style={{
```
Comment (djhi) [inline src/app/js/admin/upload/Upload.js]: put styles in an object with keys for each "classes". Make the react code easier to read

File: src/app/js/admin/upload/index.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import { createAction } from 'redux-actions';
+
+export const UPLOAD_FILE = 'UPLOAD_FILE';
+export const UPLOAD_FILE_ERROR = 'UPLOAD_FILE_ERROR';
+export const UPLOAD_FILE_PENDING = 'UPLOAD_FILE_PENDING';
+export const UPLOAD_FILE_SUCCESS = 'UPLOAD_FILE_SUCCESS';
+
+export const uploadFile = createAction(UPLOAD_FILE);
+export const uploadFilePending = createAction(UPLOAD_FILE_PENDING);
```
Comment (djhi) [inline src/app/js/admin/upload/index.js]: You don't need pending. Just use `UPLOAD_FILE` to set the loading state in the reducer

**Review by djhi — CHANGES_REQUESTED** [review]

File: src/app/js/admin/upload/index.js (hunk 9 above)
Comment (ThieryMichel) [reply src/app/js/admin/upload/index.js]: I use it to display a spinner now.

File: src/app/js/admin/parsing/ParsingResult.js [hunk 10]
```diff
@@ -101,22 +92,29 @@ export class ParsingResultComponent extends Component {
...
-    failedLines: state.parsing.failedLines,
+    failedLines: state.parsing.errors,
     loadingParsingResult: state.parsing.loading,
     totalLoadedLines: state.parsing.totalLoadedLines,
+    totalFailedLines: state.parsing.totalFailedLines,
     totalParsedLines: state.parsing.totalParsedLines,
 });
 
+const mapDispatchToProps = dispatch =>
+bindActionCreators({
```
Comment (djhi) [inline src/app/js/admin/parsing/ParsingResult.js]: why the `bindActionCreators` ? Just pass an object

File: src/app/js/admin/upload/Upload.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+import { connect } from 'react-redux';
+import RaisedButton from 'material-ui/RaisedButton';
+import ArchiveIcon from 'material-ui/svg-icons/content/archive';
+import Alert from '../../lib/Alert';
+
+import { uploadFile } from './';
+
+export const UploadComponent = ({ onFileLoad, error }) => (
+    <div
+        style={{
```
Comment (djhi) [inline src/app/js/admin/upload/Upload.js]: extract those styles outside the component in the same file with a key for each "classes", it'll make the jsx more readable

File: src/app/js/admin/upload/index.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
+import { createAction, handleActions } from 'redux-actions';
+
+export const UPLOAD_FILE = 'UPLOAD_FILE';
+export const UPLOAD_FILE_ERROR = 'UPLOAD_FILE_ERROR';
+export const UPLOAD_FILE_PENDING = 'UPLOAD_FILE_PENDING';
```
Comment (djhi) [inline src/app/js/admin/upload/index.js]: You don't need pending, set the status to 'pending' when `UPLOAD_FILE` is triggered.

**Review by djhi — CHANGES_REQUESTED** [review]

**Comment by djhi** [comment]
Also can you clean the commits ?

File: src/app/js/admin/upload/Upload.js [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+        </RaisedButton>
+    </div>
+);
+
+UploadComponent.propTypes = {
+    error: PropTypes.string.isRequired,
+    onFileLoad: PropTypes.func.isRequired,
+};
+
+const mapsStateToProps = state => ({
```
Comment (djhi) [inline src/app/js/admin/upload/Upload.js]:
```js
const mapsStateToProps = ({ upload }) => upload;
```

**Review by djhi — CHANGES_REQUESTED** [review]

File: src/app/js/admin/upload/uploadFileSaga.js [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+            file: call(loadFile, action.payload, token),
+            cancel: take([LOCATION_CHANGE]),
+        });
+        console.log({ file });
+        if (cancel) {
+            return;
+        }
+        yield put(uploadFileSuccess(file));
+    } catch (error) {
+        yield put(uploadFileError(error.message));
```
Comment (djhi) [inline src/app/js/admin/upload/uploadFileSaga.js]: pass the error directly to be FSA compliant and eventually only use the message in the reducer

**Review by djhi — CHANGES_REQUESTED** [review]

File: src/app/js/admin/upload/uploadFileSaga.js (hunk 14 above)
Comment (djhi) [reply src/app/js/admin/upload/uploadFileSaga.js]: Not doing so prevent us to use a saga or middleware for global error handling

File: src/app/js/admin/upload/Upload.js [hunk 15]
```diff
@@ -57,8 +57,8 @@ UploadComponent.propTypes = {
     onFileLoad: PropTypes.func.isRequired,
 };
 
-const mapsStateToProps = state => ({
-    ...state.upload,
+const mapsStateToProps = ({ upload }) => ({
+    ...upload,
```
Comment (djhi) [inline src/app/js/admin/upload/Upload.js]: you don't need `...`, just return `upload` directly

**Review by djhi — CHANGES_REQUESTED** [review]

**Review by djhi — APPROVED** [review]

**Pull request merged**